dali/python/nvidia/dali/plugin/pytorch.py

```diff
@@ -413,6 +413,287 @@ def __init__(self,
+                if type(out) is TensorListGPU:
+                    if not torch_gpu_device:
+                        torch_gpu_device = torch.device('cuda', dev_id)
+                    category_device[category] = torch_gpu_device
+                else:
+                    category_device[category] = torch_cpu_device
+
+            pyt_tensors = dict()
+            for j, category in enumerate(self.output_map):
+                if self._outputs_types is None or \
+                   self._outputs_types[j] == DALIGluonIterator.DENSE_TAG:
+                    pyt_tensors[category] = torch.empty(category_shapes[category],
+                                                        dtype=category_torch_type[category],
+                                                        device=category_device[category])
+                else:
+                    pyt_tensors[category] = [
+                       torch.empty(shape,
+                                   dtype=category_torch_type[category],
+                                   device=category_device[category]) 
+                       for shape in category_shapes[category]
+                    ]
+
+            data_batches[i] = pyt_tensors
+
+            # Copy data from DALI Tensors to torch tensors
+            for j, (category, tensor) in enumerate(category_tensors.items()):
+                if self._outputs_types is None or \
+                   self._outputs_types[j] == DALIGluonIterator.DENSE_TAG:
+                    if isinstance(tensor, (TensorGPU, TensorListGPU)):
+                        # Using same cuda_stream used by torch.zeros to set the memory
+                        stream = torch.cuda.current_stream(device=pyt_tensors[category].device)
+                        feed_ndarray(tensor, pyt_tensors[category], cuda_stream=stream)
+                    else:
+                        feed_ndarray(tensor, pyt_tensors[category])
+                else:
+                    for j, single_tensor in enumerate(tensor):
+                        if isinstance(tensor, (TensorGPU, TensorListGPU)):
+                            # Using same cuda_stream used by torch.zeros to set the memory
+                            stream = torch.cuda.current_stream(device=pyt_tensors[category][j].device)
+                            feed_ndarray(single_tensor, pyt_tensors[category][j], cuda_stream=stream)
+                        else:
+                            feed_ndarray(single_tensor, pyt_tensors[category][j])
+                    
+                    values = torch.hstack(pyt_tensors[category])
+                    
+                    indices = [[(i, j) for j in range(shape[0])] for i, shape in enumerate(category_shapes[category])]
+                    indices = [indice for el_indices in indices for indice in el_indices]
+                    indices = torch.LongTensor(indices, device=values.device)
+
+                    pyt_tensors[category] = torch.sparse_coo_tensor(indices.T, values)
```

Comment:
I'm not sure about this, because this is equivalent (as I far as I understand how it works) to padding all tensors with 0 to a common shape and then joining it into one. While the user may want to get instead of a single tensor a list of tensors with their shapes preserved.

Reply:
It's a bit more efficient than padding since we use a sparse `PyTorch` tensor and create it by flattening all values and creating indices of batch elements and subelements to which the values correspond. However, if the Python list of tensors with preserved shapes is more in the spirit of this change and expected behaviour, we can do it so.

Reply:
@awolant @klecki what do you think?

Reply:
I can think of use cases for both PyTorch sparse COO tensor and the list of tensors. 
What I would suggest is to make this an option for the user. It would be possible to construct sparse tensor from the list if needed but since the code for doing that is already here why not have both?
One last thing is that this option should be extendable, so for example what you have should mention `COO` since there are other variants (hybrid etc.) that we might add in the future.

Reply:
If we'd like to support multiple sparse formats then my suggestion would be to expand the list of available tags to include: `SPARSE_LIST_TAG` (for the list of tensors - same behaviour as `SPARSE_TAG` in MXNet), `SPARSE_COO_TAG`, `SPARSE_CSR_TAG` etc. - for all sparse implementations. What do you think @awolant , @JanuszL ?

Reply:
Sounds good. I think you can now add `SPARSE_COO_TAG` and state in the docs `SPARSE_TAG` defaults to `SPARSE_COO_TAG` and in the feature more formats could be added.

Reply:
I've added `SPARSE_COO_TAG` and also `SPARSE_LIST_TAG` for the list of tensors since we already have this list before building a sparse tensor. I've updated the docs and stated that for now only these two sparse formats are supported.

Reply:
Looks good. Thank you for that.